### PR TITLE
[SYCL][Fusion] Add kernel fusion extension API

### PR DESCRIPTION
**Pull request by sommerlukas**
This is the first patch in a series of patches to add an implementation of the [kernel fusion extension](https://github.com/intel/llvm/pull/7098). We have split the implementation into multiple patches to make them more easy to review.

This first patch introduces the user-facing API discussed in the [extension proposal](https://github.com/intel/llvm/pull/7098). It does not yet add any fusion functionality, just the mere API and SYCL properties. Calls to the API will throw an error until the remaining functionality lands with the following patches. 

The design document for the overall implementation of kernel fusion can be found [here](https://github.com/intel/llvm/pull/7204).

Signed-off-by: Lukas Sommer <lukas.sommer@codeplay.com>

### CODE REVIEW

File: sycl/CMakeLists.txt
Comment (sommerlukas) [reply sycl/CMakeLists.txt:148]: A future PR will add support for fusion in `buildbot/configure.py`, which would then set the value for this option. 

Users using `buildbot/configure.py` would therefore not need to set this option manually.

Comment (pvchupin) [reply sycl/CMakeLists.txt:148]: I guess it can be documented at https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#build-dpc-toolchain

Comment (steffenlarsen) [reply sycl/CMakeLists.txt:148]: In that case, I am okay with waiting with documentation for it until it has an option in `buildbot/configure.py`. I'll let @pvchupin have the last say in that though.

File: sycl/include/sycl/ext/codeplay/experimental/fusion_wrapper.hpp
Comment (steffenlarsen) [inline sycl/include/sycl/ext/codeplay/experimental/fusion_wrapper.hpp:87]: It is currently experimental so we can break it in the future. However, it would be better for `fusion_wrapper` to have its own `fusion_wrapper_impl` class so it can be extended without breaking ABI. I am okay to have that in a follow-up patch if that unblocks further work.

File: sycl/test/abi/sycl_symbols_linux.dump
Comment (steffenlarsen) [inline sycl/test/abi/sycl_symbols_linux.dump:4043]: Note that CI does not run the Windows variant of this test, but post-commit does. As such, when this is merged post-commit will likely start complaining. If you have a windows machine you can run tests with, it might be a good idea to give this patch a run-through with that. If not, please add the missing symbols ASAP after this has been merged.

**Review by steffenlarsen — APPROVED** [review]
LGTM! As mentioned in the comment, I am okay with having the `kernel_funsion_impl` class in a follow-up patch.

File: sycl/include/sycl/ext/codeplay/experimental/fusion_wrapper.hpp
Comment (sommerlukas) [reply sycl/include/sycl/ext/codeplay/experimental/fusion_wrapper.hpp]: Thanks for the feedback, I've added the `fusion_wrapper_impl` as part of this PR, see d2ff9ab1f459bc75e7cc55744152338d5dd84843.

File: sycl/test/abi/sycl_symbols_linux.dump
Comment (sommerlukas) [reply sycl/test/abi/sycl_symbols_linux.dump:4043]: Thanks for pointing that out @steffenlarsen!

We would like to avoid breaking the post-commit pipeline, so I've updated the corresponding Windows symbol dump/test in 11349a7bba3738fbdf5640335872e5806365c227 as part of this PR.

Comment (steffenlarsen) [reply sycl/test/abi/sycl_symbols_linux.dump:4043]: Awesome, thanks!

File: sycl/include/sycl/ext/codeplay/experimental/fusion_wrapper.hpp
Comment (steffenlarsen) [reply sycl/include/sycl/ext/codeplay/experimental/fusion_wrapper.hpp]: Great, thank you! 😄

**Review by steffenlarsen — APPROVED** [review]
Looks good! 🚀

**Comment by sommerlukas** [comment]
Thanks for the approval @steffenlarsen. Can you or @pvchupin please merge this PR, I do not have the necessary permissions to merge.

**Pull request merged**
